File: src/C++/SOM.cpp
#include "SOM.h"

/* 
	Construct untrained SOM with given lattice width and height
*/
SOM::SOM(unsigned int width, unsigned int height)
{
	this->_width = width;
	this->_height = height;
}

/*
	Construct SOM from a saved SOM width, height, and set of weights
*/
SOM::SOM(std::istream &in) {
	this->load_weights(in);
}

/*
	Train the SOM using a set of training data over a given number of epochs with a given learning rate
*/
void SOM::train_data(double *trainData, unsigned int num_examples, unsigned int dimensions, int epochs, double initial_learning_rate)
{
	this->_dimensions = dimensions;
	// Normalize data (to be within 0 to 1)
	normalizeData(trainData, num_examples);

	// Randomly initialize codebook
	this->_weights = (double *)malloc(_width * _height * _dimensions * sizeof(double));
	for (int i = 0; i < _width; i++) {
		for (int j = 0; j < _height; j++) {
			for (int d = 0; d < _dimensions; d++) {
				this->_weights[calcIndex(i,j,d)] = randWeight();
			}
		}
	}

	// Calc initial map radius
	double initial_map_radius = _width < _height ? ((double)_width) / 2.0 : ((double)_height) / 2.0;

	// Find BMUs for every input instance
	// D = X_sq - 2X^TM + M_sq
	// D (xdn * nn)
	double* D = (double *)malloc(num_examples * _width * _height);

	// Calc m_sq
	double* m_sq = (double *)malloc(_width * _height);
	SqDists(this->_weights, _width * _height, _dimensions, m_sq);

	// Calc x_sq
	double* x_sq = (double *)malloc(num_examples);
	SqDists(trainData, num_examples, _dimensions, x_sq);

	for (int i = 0; i < num_examples; i++) {
		for (int j = 0; j < _width; j++) {
			for (int k = 0; k < _height; k++) {
				// Calc x^Tm
				double xm = 0;
				for (int d = 0; d < _dimensions; d++) {
					xm += trainData[i *_dimensions + d] * this->_weights[(j * _height + k) * _dimensions + d];
				}
				// Combine all
				D[(i * _width + j) * _height + k] = x_sq[i] - 2 * xm + m_sq[j * _height + k];
			}
		}
	}
	free (m_sq);
	free (x_sq);

	// BMU index of each training instance
	int* BMUs = (int *)malloc(num_examples);
	for (int j = 0; j < num_examples; j++) {
		BMUs[j] = 0;
		for (int i = 1; i < _width * _height; i++) {
			if (D[j * _width * _height + i] < D[j * _width * _height + BMUs[j]]) {
				BMUs[j] = i;
			}
		}
	}

	// Calc N for each node
	int* N = (int *)malloc(_width * _height);
	for (int i = 0; i < _width * _height; i++) {
		N[i] = 0;
	}
	for (int j = 0; j < num_examples; j++) {
		N[BMUs[j]]++;
	}

	// Calc gaussian function 
	// (num_examples x num nodes)
	double* H = (double *)malloc(num_examples * _width * _height);
	for (int j = 0; j < num_examples; j++) {
		for (int i = 0; i < _width * _height; i++) {
			H[j*_width*_height + i] = h(j, i, initial_map_radius, initial_map_radius, BMUs);
		}
	}

	// Update codebook
	for (int i = 0; i < _width * _height; i++) {
		// (H^T(num nodes x num_examples) * X(num_examples x _dimensions)) * n_j (scalar) = numerators (num nodes x _dimensions) ?
		// For denominators, H is left-multiplied by an num_examples dimensional vector of ones
	}

	free(H);
	free(N);
	free(D);
	free(BMUs);
}

/*
	Save the width and height of the SOM followed by the weights for each node with a different node's weights on every line
*/
void SOM::save_weights(std::ostream &out)
{
	out << this->_width << " " << this->_height << std::endl;
	for (int i = 0; i < this->_width; i++)
	{
		for (int j = 0; j < this->_height; j++)
		{
			for (int k = 0; k < this->_dimensions; k++) {
				if (k != 0) {
					out << " ";
				}
				out << this->_weights[calcIndex(i,j,k)];
			}
			out << std::endl;
		}
	}
}

/*
	Load a trained SOM that was saved using the same algorithm as save_weights from an input stream
*/
void SOM::load_weights(std::istream &in)
{
	// Load SOM dimensions first
	in >> this->_width >> this->_height;

	// Read first line of matrix to get the dimensionality of weights
	this->_dimensions = 0;
	std::string line;
	std::getline(in, line);
	std::getline(in, line);
	std::stringstream ss(line);
	std::vector<double> line1;
	double temp;
	while (ss >> temp) {
		this->_dimensions++;
		line1.push_back(temp);
	}

	// Put first line of matrix into an array in the 3d weights array
	this->_weights = new double[_width * _height * _dimensions];
	for (int k = 0; k < this->_dimensions; k++) {
		_weights[calcIndex(0,0,_dimensions - k - 1)] = line1.back();
		line1.pop_back();
	}

	// Read the rest of the 3d array in
	for (int i = 0; i < this->_width; i++) {
		for (int j = (i == 0 ? 1 : 0); j < this->_height; j++) {
			for (int k = 0; k < _dimensions; k++) {
				in >> this->_weights[calcIndex(i,j,k)];
			}
		}
	}
}

/*
	Normalizes given data to be between 0 and 1 for each feature
*/
void SOM::normalizeData(double *trainData, int num_examples)
{
	// Find the max and min value for each feature then use it to normalize the feature
	this->_featureMaxes = new double[this->_dimensions];
	this->_featureMins = new double[this->_dimensions];
	for (int d = 0; d < this->_dimensions; d++)
	{
		this->_featureMaxes[d] = -std::numeric_limits<double>::max();
		this->_featureMins[d] = std::numeric_limits<double>::max();
		for (int i = 0; i < num_examples; i++)
		{
			if (trainData[i*_dimensions + d] > this->_featureMaxes[d]) {
				this->_featureMaxes[d] = trainData[i*_dimensions + d];
			}
			if (trainData[i*_dimensions + d] < this->_featureMins[d]) {
				this->_featureMins[d] = trainData[i*_dimensions + d];
			}
		}
		for (int i = 0; i < num_examples; i++) {
			trainData[i*_dimensions + d] = (trainData[i*_dimensions + d] - this->_featureMins[d])/(this->_featureMaxes[d]-this->_featureMins[d]);
		}
	}
}

/*
	Update a node's weights to better match a given example
*/
void SOM::updateNodeWeights(int x, int y, double* example, double learning_rate, double influence) {
	for (int d = 0; d < this->_dimensions; d++)
	{
		this->_weights[calcIndex(x,y,d)] += influence * learning_rate * (example[d] - this->_weights[calcIndex(x,y,d)]);
	}
}

/*
	Generate a vector of size numFeatures
*/
double SOM::randWeight()
{
	return (double)rand() / (RAND_MAX);
}

int SOM::calcIndex(int x, int y, int d) {
	return (x*_height + y)*_dimensions + d;
}

/*
	Calculates the euclidean distance between two vectors
*/
double SOM::EucDist(double* v1, double* v2) {
	double total = 0.0;
	for (int i = 0; i < this->_dimensions; i++) {
		total += (v1[i] - v2[i])*(v1[i] - v2[i]);
	}
	return sqrt(total);
}

void SOM::SqDists(double* m, int loop, int dim, double* output) {
	for (int i = 0; i < loop; i++) {
		output[i] = 0;
		for (int d = 0; d < dim; d++) {
			output[i] += m[i * d + d] * m[i * d + d]; 
		}
	}
}

double SOM::h(int j, int i, double initial_radius, double radius, int* BMUs) {
	int i_y = i % _height;
	int i_x = (i - i_y) / _height;

	// Get BMU coord
	int j_y = BMUs[j] % _height;
	int j_x = (BMUs[j] - j_y) / _height;

	return initial_radius * exp(-(double)((j_x - i_x) * (j_x - i_x) + (j_y - i_y) * (j_y - i_y))/(radius * radius));
}

File: src/Python/som.py
import argparse
import numpy as np
import math
import random
import matplotlib.pyplot as plt
import matplotlib.cm as cm
import time
import functools

l_0 = 0.1

def euc(x):
	s = 0
	for i in x:
		s += i ** 2
	return s

class SOM:
	def __init__(self, width, height, squareneurons=False):
		# 'Generate' node lattice
		self._width = width
		self._height = height
		self._hex = not squareneurons
	
	def train(self, train_data, iterations):
		# Normalize data to be between 0 and 1
		train_data = train_data / train_data.max(axis=0)

		# Randomly initialize weights
		self._weights = np.random.rand(self._width, self._height, train_data.shape[1])

		# Initialize NxNx2 matrix where positions[i,j] = [i,j] for the node distance calculations
		positions = np.array(np.meshgrid(np.arange(self._width),np.arange(self._height))).T.reshape(self._width,self._height,2)

		# Initialize map radius
		map_radius = float(max(self._width,self._height) / 2)

		# Initialize time constant
		tc = float(iterations)/math.log(map_radius)

		# Repeat N times
		for ts in range(iterations):
			# Current learning rate
			l_t = l_0 * math.exp(-float(ts)/tc)

			for example in train_data:
				# Calculate the euclidean distance between the weights and the example
				weight_sq_dists = np.apply_along_axis(euc, 2, np.subtract(self._weights, example))
				# Select the BMU
				bmu_index = np.unravel_index(np.argmin(weight_sq_dists, axis=None), weight_sq_dists.shape)

				# Calculate radius of the neighborhood radius of the BMU (start large, diminish each time step) (any nodes within the raidus are inside the BMU's neighborhood)
				neighborhood_radius = map_radius * math.exp(-(float(ts))/tc)

				# Adjust each node's weights (to make more like the chosen vector) (adjust by factor of closeness to BMU)
				# Vectorized function to calculate the theta that partially determines the amount each weight is modified to match the BMU (based on the squared distance to bmu)
				theta_lambda_v = np.vectorize(lambda x: math.exp(-(x)/(2*(neighborhood_radius ** 2))))
				
				# Calculate the distance of each position to the BMU
				dists = np.apply_along_axis((lambda x: x[0] ** 2 + x[1] ** 2), 2, np.subtract(positions, np.array(bmu_index)))
				# Perform theta calculation for each weight
				thetas = theta_lambda_v(dists)

				# Update weights
				self._weights = self._weights + l_t * thetas[:,:,np.newaxis] * (example - self._weights)

	def saveWeights(self, filename):
		np.savetxt(filename, self._weights.reshape(-1,self._weights.shape[2]), header=str(self._width) + ' ' + str(self._height), comments='')

	def loadWeights(self, filename):
		dim = np.loadtxt(filename, max_rows=1)
		self._width = int(dim[0])
		self._height = int(dim[1])

		temp = np.loadtxt(filename, skiprows=1)
		self._weights = np.reshape(temp, (self._width,self._height,temp.shape[1]))

	def displayTopology(self):
		plt.figure()
		plt.imshow(np.average(self._weights, axis=2), cmap=cm.get_cmap(name="RdBu"), interpolation='bicubic')
		plt.title("Topology")

	def displayInputPlanes(self):
		for k in range(self._weights.shape[2]):
			plt.figure()
			plt.imshow(self._weights[:,:,k], cmap=cm.get_cmap(name="RdBu"), interpolation='bicubic')
			plt.title("Input plane " + str(k+1))
	
	def displayColor(self):
		plt.figure()
		plt.imshow(self._weights, interpolation='bicubic')
	
	def displayUMatrix(self):
		# Step 1: Calculate the distance between each node and its neighbors (for square node maps)
		umatrix = np.empty(self._weights.shape[:-1])
		for row in range(self._weights.shape[0]):
			for col in range(self._weights.shape[1]):
				dist = 0
				n_neighbors = 0
				# Up
				if row != self._weights.shape[0] - 1:
					dist += math.sqrt(euc(np.subtract(self._weights[row][col], self._weights[row+1][col])))
					n_neighbors += 1
				# Right
				if col != self._weights.shape[1] - 1:
					dist += math.sqrt(euc(np.subtract(self._weights[row][col], self._weights[row][col+1])))
					n_neighbors += 1
				# Down
				if row != 0:
					dist += math.sqrt(euc(np.subtract(self._weights[row][col], self._weights[row-1][col])))
					n_neighbors += 1
				# Left
				if col != 0:
					dist += math.sqrt(euc(np.subtract(self._weights[row][col], self._weights[row][col-1])))
					n_neighbors += 1
				dist /= n_neighbors
				umatrix[row][col] = dist
		plt.figure()
		plt.imshow(umatrix, cmap=cm.get_cmap(name="RdBu"), interpolation='bicubic')

def getArguments():
	parser = argparse.ArgumentParser(description="Generates a SOM from a given file source or loads a pretrained SOM from a file. Can also display SOMs using a variety of display methods.")

	# SOM dimensions
	parser.add_argument('width', type=int, nargs='?', default=0, help='The width of the SOM.')
	parser.add_argument('height', type=int, nargs='?', default=0, help='The height of the SOM.')
	# Training data (determines SOM dimensionality)
	parser.add_argument('source', nargs='?',type=argparse.FileType('r'), help='The source data for the SOM to be trained on.')
	
	# Number of training iterations
	parser.add_argument('-N', '--trainingiterations', type=int, default=100, help='The number of iterations to do in training.')
	# Output file for weights
	parser.add_argument('-o', '--out', type=argparse.FileType('w'), help='The output file for the SOM\'s node weights.')
	# Input file for weights
	parser.add_argument('-i', '--input', type=argparse.FileType('r'), help='The input file for the SOM\'s node weights (does not retrain loaded map).')

	# Square/Hex neurons
	parser.add_argument('-s', '--squareneurons', type=bool, nargs='?', const=True, default=False, help='Whether the neurons in the map should be square (otherwise hexagonal).')
	
	# Display methods
	parser.add_argument('-d', '--display', action='append', nargs='?', type=str, const='topology', default=[], help='The display method to be used on the SOM. Current usable display methods: topology, input-planes, color, u-matrix. If no argument after flag then defaults to topology') 

	args = parser.parse_args()
	for dm in args.display:
		if dm not in ['topology','input-planes','color','u-matrix']:
			print('Invalid display method \'' + dm + '\'')
			return None

	return args

if __name__ == '__main__':
	args = getArguments()

	if args != None:
		if args.input != None:
			# Load trained SOM
			s = SOM(0,0)
			s.loadWeights(args.input.name)
		else:
			# Pull data from file
			data = np.genfromtxt(args.source.name)
			s = SOM(args.width, args.height, squareneurons=args.squareneurons)

			t0 = time.time()
			# Train SOM
			s.train(data, iterations=args.trainingiterations)
			t1 = time.time()

			print('Total training time:', t1-t0, flush=True)

		if args.out != None:
			s.saveWeights(args.out.name)

		if len(args.display) > 0:
			for dm in args.display:
				if dm == 'topology':
					s.displayTopology()
				elif dm == 'input-planes':
					s.displayInputPlanes()
				elif dm == 'color':
					if s._weights.shape[2] != 3:
						print('Display method color invalid for dimensions not equal to 3')
					else:
						s.displayColor()
				elif dm == 'u-matrix':
					s.displayUMatrix()
			plt.show()